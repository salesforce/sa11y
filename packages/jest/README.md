# `jest`

Accessibility matcher for [Jest](https://jestjs.io)

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->


- [Setup](#setup)
  - [Project level](#project-level)
  - [Test module level](#test-module-level)
- [Usage](#usage)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Setup

The accessibility matcher helper APIs need to be registered with Jest before they can be used in tests.

### Project level

You can set up the a11y matchers once at the project level to make it available to all the Jest tests in the project.
For an example look at the [Integration tests](../test-integration/README.md).

-   Add a Jest setup file (e.g. `jest-setup.js`) and add the following code that registers the a11y matchers

```javascript
const { registerA11yMatchers } = require('@sa11y/jest');
registerA11yMatchers();
```

-   Add/Modify Jest config at project root to invoke the Jest setup file as setup above.
    In the `jest.config.js` at the root of your project, add:

```javascript
module.exports = {
    setupFilesAfterEnv: ['<rootDir>/jest-setup.js'],
};
```

### Test module level

Invoke `registerA11yMatchers` before using the accessibility matchers in the tests e.g.

```javascript
import { registerA11yMatchers } from '@sa11y/jest';

beforeAll(() => {
    registerA11yMatchers();
});
```

## Usage

-   **WARNING**: `toBeAccessible` **must** be invoked with `await` or the equivalent supported async method in your env.
    -   Not invoking it async would result in incorrect results
    -   e.g. no assertions even when page is not accessible
-   `toBeAccessible` jest matcher can be invoked on a `document` (JSDOM) or an HTML element
    to check for accessibility

```javascript
import { recommended } from '@sa11y/preset-rules';
import { registerA11yMatchers } from '@sa11y/jest';

beforeAll(() => {
    registerA11yMatchers();
});

it('should be accessible', async () => {
    // Setup DOM to be tested for accessibility
    //...

    // assert that DOM is accessible (using extended preset-rule)
    await expect(document).toBeAccessible();

    // Can be used to test accessibility of a specific HTML element
    const elem = document.getElementById('foo');
    await expect(elem).toBeAccessible();

    // use recommended preset-rule
    await expect(document).toBeAccessible(recommended);
});
```
