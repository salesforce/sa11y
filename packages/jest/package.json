{
    "name": "@sa11y/jest",
    "version": "0.1.3-beta.0",
    "description": "Accessibility testing matcher for Jest",
    "license": "BSD-3-Clause",
    "homepage": "https://github.com/salesforce/sa11y/tree/master/packages/jest#readme",
    "repository": {
        "type": "git",
        "url": "https://github.com/salesforce/sa11y.git",
        "directory": "packages/jest"
    },
    "keywords": [
        "accessibility",
        "automated testing",
        "axe",
        "jest",
        "matcher"
    ],
    "main": "dist/index.js",
    "types": "dist/index.d.ts",
    "files": [
        "dist/**/*.js"
    ],
    "dependencies": {
        "@sa11y/assert": "0.1.3-beta.0",
        "@sa11y/format": "0.2.1-beta.0",
        "@sa11y/preset-rules": "0.2.0-beta.0",
        "jest-matcher-utils": "^26.1.0"
    },
    "peerDependencies": {
        "jest": ">=26.0.0"
    },
    "devDependencies": {
        "@sa11y/common": "0.2.0-beta.0",
        "@sa11y/test-utils": "0.2.0-beta.0"
    },
    "publishConfig": {
        "access": "public"
    }
}
