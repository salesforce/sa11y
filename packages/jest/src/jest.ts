/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { assertAccessible } from '@sa11y/assert';
import { A11yConfig, extended } from '@sa11y/preset-rules';
import { matcherHint } from 'jest-matcher-utils';

export const matcherHintMsg = `expected document to have no accessibility violations but found following issues: `;

// Type def for custom jest a11y matcher toBeAccessible
// Ref: https://jestjs.io/docs/en/expect.html#expectextendmatchers
declare global {
    // eslint-disable-next-line @typescript-eslint/no-namespace
    namespace jest {
        interface Matchers<R> {
            toBeAccessible(config?: A11yConfig): Promise<CustomMatcherResult>;
            toBeAccessibleWith(config: A11yConfig): Promise<CustomMatcherResult>;
        }
    }
}

/**
 * Jest expect matcher to check DOM for accessibility issues
 * @param receivedDom - DOM to be tested for accessibility
 * @param config - A11yConfig to be used to test for accessibility. Defaults to extended.
 */
export async function toBeAccessible(
    receivedDom: Document = document,
    config: A11yConfig = extended
): Promise<jest.CustomMatcherResult> {
    let isAccessible = true;
    let a11yViolations = '';

    try {
        await assertAccessible(receivedDom, config);
    } catch (e) {
        isAccessible = false;
        a11yViolations = e;
    }
    return {
        pass: isAccessible,
        // Display assertion for the report when a test fails
        message: (): string => matcherHint(`${toBeAccessible.name}: ${matcherHintMsg} \n\n ${a11yViolations}`),
    };
}

/**
 * Jest expect matcher to check DOM for accessibility issues with a `@sa11y/preset-rule`
 * @param receivedDom - DOM to be tested for accessibility
 * @param config - A11yConfig to be used to test for accessibility.
 */
export function toBeAccessibleWith(
    receivedDom: Document = document,
    config: A11yConfig
): Promise<jest.CustomMatcherResult> {
    return toBeAccessible(receivedDom, config);
}
