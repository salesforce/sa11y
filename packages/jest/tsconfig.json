{
    "extends": "../../tsconfig.common.json",
    "compilerOptions": {
        "rootDir": "src",
        "outDir": "dist"
    },
    "include": ["src"],
    "references": [{ "path": "../preset-rules" }, { "path": "../assert" }, { "path": "../test-utils" }]
}
