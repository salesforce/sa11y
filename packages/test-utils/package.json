{
    "name": "@sa11y/test-utils",
    "version": "0.2.0-beta.0",
    "private": true,
    "description": "Private package providing test utilities for @sa11y packages",
    "license": "BSD-3-Clause",
    "homepage": "https://github.com/salesforce/sa11y/tree/master/packages/test-utils#readme",
    "repository": {
        "type": "git",
        "url": "https://github.com/salesforce/sa11y.git",
        "directory": "packages/test-utils"
    },
    "main": "dist/index.js",
    "types": "dist/index.d.ts",
    "files": [
        "dist/**/*.js"
    ],
    "dependencies": {
        "@sa11y/common": "0.2.0-beta.0"
    }
}
