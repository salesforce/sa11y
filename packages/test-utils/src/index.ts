/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export {
    audioURL,
    exceptionList,
    domWithA11yIssues,
    domWithDescendancyA11yIssues,
    domWithA11yIssuesBodyID,
    domWithNoA11yIssues,
    domWithNoA11yIssuesChildCount,
    domWithVisualA11yIssues,
    htmlFileWithA11yIssues,
    htmlFileWithNoA11yIssues,
    htmlFileWithVisualA11yIssues,
    a11yIssuesCount,
    a11yIssuesCountFiltered,
    shadowDomID,
    videoURL,
    customRulesFilePath,
    domWithA11yCustomIssues,
} from './test-data';
export { beforeEachSetup, cartesianProduct } from './utils';
