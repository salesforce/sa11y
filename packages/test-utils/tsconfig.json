{
    "extends": "../../tsconfig.common.json",
    "compilerOptions": {
        "rootDir": "src",
        "outDir": "dist"
    },
    "include": ["src"],
    "references": [{ "path": "../common" }]
}
