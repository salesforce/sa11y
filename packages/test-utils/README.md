# `@sa11y/test-utils`

Private package providing test utilities for @sa11y packages

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->


- [Usage](#usage)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Usage

```javascript
import { beforeEachSetup, domWithA11yIssues, domWithNoA11yIssues } from '@sa11y/test-utils';

import { registerSa11yMatcher } from '@sa11y/jest';

beforeAll(() => {
    registerSa11yMatcher();
});

beforeEach(() => {
    beforeEachSetup();
});

describe('...', () => {
    it('...', async () => {
        document.body.innerHTML = domWithNoA11yIssues;
        await expect(document).toBeAccessible();

        document.body.innerHTML = domWithA11yIssues;
        await expect(document).not.toBeAccessible();
    });
});
```
