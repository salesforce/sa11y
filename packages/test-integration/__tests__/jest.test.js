/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { beforeEachSetup, domWithA11yIssues, domWithNoA11yIssues, domWithVisualA11yIssues } from '@sa11y/test-utils';
import { registerSa11yMatcher } from '@sa11y/jest';
import { expect } from '@jest/globals';

beforeAll(registerSa11yMatcher);

beforeEach(beforeEachSetup);

describe('integration test @sa11y/jest', () => {
    /* eslint-disable @typescript-eslint/no-unsafe-call */
    it('should have a11y matchers working with setup in jest.config.js', async () => {
        expect(expect.toBeAccessible).toBeDefined();
        document.body.innerHTML = domWithNoA11yIssues;
        await expect(document).toBeAccessible();
    });

    it('should throw error for inaccessible dom', async () => {
        document.body.innerHTML = domWithA11yIssues;
        await expect(expect(document).toBeAccessible()).rejects.toThrow();
    });

    it('will not throw error for audio video color-contrast', async () => {
        document.body.innerHTML = domWithVisualA11yIssues;
        // Even though the dom has a11y issues w.r.t color contrast and audio/video
        //  elements etc, Jest/JSDOM will not able to detect them
        await expect(document).toBeAccessible();
    });
    /* eslint-enable @typescript-eslint/no-unsafe-call */
});
