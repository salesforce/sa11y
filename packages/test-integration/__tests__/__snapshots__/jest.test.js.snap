// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration test @sa11y/jest should throw error for inaccessible dom 1`] = `
"[2mexpect([22m[31m3 issues[39m[2m)toBeAccessible: expected document to have no accessibility violations but found 3 issues: [22m
[2m[22m
[2m [31m\\"â­• (document-title) Documents must have <title> element to aid in navigation: html\\"[39m[22m
[2m	ðŸ”— Help URL: https://dequeuniversity.com/rules/axe/4.3/document-title[22m
[2m[22m
[2m[31m\\"â­• (link-name) Links must have discernible text: a\\"[39m[22m
[2m	ðŸ”— Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name[22m
[2m[22m
[2m[31m\\"â­• (page-has-heading-one) Page should contain a level-one heading: html\\"[39m[22m
[2m	ðŸ”— Help URL: https://dequeuniversity.com/rules/axe/4.3/page-has-heading-one([22m[32m0 issues[39m[2m)[22m"
`;
